pandas/tests/indexes/test_base.py

```diff
@@ -305,6 +305,30 @@ def test_constructor_simple_new(self):
         result = idx._simple_new(idx, 'obj')
         tm.assert_index_equal(result, idx)
 
+    def test_constructor_names(self):
+        # test both `name` and `names` provided
+        with pytest.raises(TypeError):
```

Comment:
can you have these first two assertions in a separate test that uses ``pandas.util.testing.all_index_generator``, and passes the ``name`` to exhaustively check all the index types (do this in a parameterization)

Reply:
Please, correct me if I'm wrong, but don't you think it's a bit redundant? I mean I can't use `name` or `names` for `all_index_generator` to check constructors of different index types. All I can do is to create a new `Index` with each of the generated indexes passing `names` and/or `name`. In this case it doesn't really matter what type of index is going to be created as all the checks are made at the very beginning of `Index` constructor (before objects creation) and they should guarantee that `name` and `names` are valid for any case.

Reply:
how is this redundant? we need to generally test passing name & names to each of the individual constructors. I suspect this actually does break some of the subclasses. its simple to modify ``all_index_generator`` to pass thru kwargs

Reply:
None of the `Index` subclasses (apart from `MultiIndex`) sets `names` for their objects. They don't even consider this argument. It shouldn't break anything. The only purpose for `names` is to create a `MultiIndex` instance (either via `Index` or `MultiIndex` constuctor).

Reply:
@toobaz , could you, please, add any remarks about the tests too?

Reply:
My remarks here are not worth much as I'm not a core dev (I can only share your frustration :-) )

Theoretically, it could be that some day somebody modified some ``Index`` subclass so that it misbehaves on ``names``. Your tests would capture that regression. Yeah, I know, that's paranoia - but also the quickest way to get your PR accepted.

This said: _if_ we end up deprecating ``names`` (as I think we should, but I don't think there ever was any discussion, so not sure how it would end up), then all of this is makes even less sense, so you _might_ try to open that discussion now.

Reply:
It's not just paranoia, I think, it's something that can just mess things up. We're talking about error cases, but `Index` subclasses don't consider `names` at all, so they won't raise any errors when passing wrong names arguments as it's expected for `Index` and `MultiIndex`. So, to make these tests work correctly, I'll have to modify all the subclasses too.
The more I think about it, the more I dislike the way things are going :\

Reply:
OK, I think it's time for #19295 ;-)

Reply:
@PoppyBagel here's the big pictures. We want to consider the consistency guarantees that pandas offers. We have a loose guaranteee that using the ``Index`` constructor or its sub-classes have the same signature, IM is an exception to this. But for example ``name`` is accepted.

We want to guaranteee this behavior *in one place* via tests. It is somewhat tested in each subclass, but not centrally. Further this locks down the behavior of *future* subclasses, and prevent inadvertant adding it back. More to the point, I am considering removing ``names`` entirely (via deprecation). We need to check for error messages for this.

This is a little bit overkill, to check that we are *not* accepting an argument, but we have lots of code and inspection can easily fail.

I don't want you to *add* the ``names`` args anywhere, If its *only* accepted in MI and Index, then we can next consider the removal implications.